pkg/kubelet/cm/helpers.go

```diff
@@ -44,3 +51,33 @@ func hardEvictionReservation(thresholds []evictionapi.Threshold, capacity v1.Res
 	}
 	return ret
 }
+
+func buildContainerMapAndRunningSetFromRuntime(ctx context.Context, runtimeService internalapi.RuntimeService) (containermap.ContainerMap, sets.String) {
+	podSandboxMap := make(map[string]string)
+	podSandboxList, _ := runtimeService.ListPodSandbox(ctx, nil)
+	for _, p := range podSandboxList {
+		podSandboxMap[p.Id] = p.Metadata.Uid
+	}
+
+	runningSet := sets.NewString()
+	containerMap := containermap.NewContainerMap()
+	containerList, _ := runtimeService.ListContainers(ctx, nil)
+	for _, c := range containerList {
+		if _, exists := podSandboxMap[c.PodSandboxId]; !exists {
+			klog.InfoS("No PodSandBox found for the container", "podSandboxId", c.PodSandboxId, "containerName", c.Metadata.Name, "containerId", c.Id)
+			continue
+		}
+		podUID := podSandboxMap[c.PodSandboxId]
+		containerMap.Add(podUID, c.Metadata.Name, c.Id)
+		if isContainerStateRunning(c.State) {
+			klog.V(4).InfoS("Container reported running", "podSandboxId", c.PodSandboxId, "podUID", podUID, "containerName", c.Metadata.Name, "containerId", c.Id)
+			runningSet.Insert(c.Id)
+		}
+	}
+
+	return containerMap, runningSet
+}
+
+func isContainerStateRunning(state runtimeapi.ContainerState) bool {
+	return state == runtimeapi.ContainerState_CONTAINER_RUNNING
+}
```

Comment:
I don' think this helper is necessary, it'a clearer to me if it is simply inlined.

That said, if we do what I suggested previously, and rack the full container object instead of just its running state, then this check would be done elsewhere anyway.

Reply:
After few days of consideration, I think we should _not_ rack the full container object:
1. I for myself I think we're doing a bit of a layering violation querying the runtime in `buildContainerMap`. Or at very least we should aim to reconsider/refactor, which is a bigger effort outside of the scope of this PR
2. I would like this fix to be easily backportable because we likely need to backport down to 1.25 (or at least enable it) so I'm aiming to keep the changes minimal.

For both these reasons I'm keeping the separate set, even though I agree is a bit uglier, and I'm gonna close https://github.com/kubernetes/kubernetes/pull/119151